docs/enhancements/boot-status-reporter.md

```diff
@@ -0,0 +1,129 @@
+---
+title: boot-status-reporter
+authors:
+  - "@nmagnezi"
+creation-date: 2022-09-22
+last-updated: 2022-09-22
+---
+
+# Assisted Boot Reporter
+
+## Summary
+
+Create a new reporting service that runs as early as possible when hosts are booted from disk to collect additional 
+information and send it to `assisted-service`.
+
+## Motivation
+
+There are multiple occurrences of clusters that fail to install at the stage where
```

Comment:
So the case here is when the host did boot successfully from the right disk, configured network, and pulled the ignition.
In this case we have the installer running on the bootstrap node that should collect the relevant logs from the rest of the nodes.
In case the problem happened in the bootstrap node, the assisted-installer-controller should collect the relevant logs.

Reply:
>In case the problem happened in the bootstrap node, the assisted-installer-controller should collect the relevant logs.

The controller can't collect anything about the bootstrap node's post reboot early boot processr

Reply:
This case is mostly relevant for sno, in case controller is not started due to some issues like kube-api/kubelet and etc
we will get logs

Reply:
I agree that this service is mostly relevant for SNO and I think the enhancement motivation, background and goals should reflect that.

>  The controller can't collect anything about the bootstrap node's post reboot early boot process 
This is something we can solve with the current implementation (copy the private key to the controller).

On the other hand, if the motivation is to have a robust mechanism to get the information about the host boot status and the logs from the host regardless of the scenario (SNO, worker stuck in configuring, host changed IP) than this service makes sense.

Reply:
when will it be even relevantl:
1. wrong boot order - wont help
2. no networking post reboot (static networking was not applied correctly) - low chances that despite that, the host will still have networking to reach the saas / hub cluster
3. wrong dns configuration - lets add better validations before triggering the installation
4. what else?

Reply:
Literally any error that prevents crio/kubelet from functioning properly (e.g. expired certs? the crio ipv4/ipv6 bug we had recently?)

Reply:
i.e. crio and kubelet and super complex and prone to failure, we want to see why and how they fail even on SNO/bootstrap nodes

Reply:
Also we can't really tell why some installations fail when it's the bootstrap node post-reboot or a SNO.

This service is a best-effort attempt at maybe finding an answer (or maybe not, depending on the failure)

Reply:
I discussed with Nir quite a bit about the motivation and I would like to write it explicitly here - any solution that involves running assisted-installer-controller is not a solution here.

Reasoning - in order to run assisted-installer-controller we need to have a working k8s cluster. In order to have a working k8s cluster we need to have a working kubelet. In order to have a working kubelet we need to have a working crio. Therefore, before booting a machine and getting to run assisted-installer-controller there is a bunch of components out of which none reports its status mid-way.

To have a robust solution we need to have something that is plugged at the beginning of this dependency chain. Thus, we need something with no dependency on crio. Thus, our solution cannot be running a container. This takes assisted-installer-controler out of the game here.

We keep discussing improving assisted-installer-controller for ages now (and it's good, it needs improvements) but none of those discussions focuses on the problem we try to solve here - installation can fail because a component of the RHCOS is dodged. We need ability to debug this, otherwise we will keep receive tickets that are unsolvable without overseeing the installation in real-time.

We don't need an ideal solution. We need a solution good enough that will tell us e.g. "machine booted with network connectivity but XYZ happened so kubelet never started". Omer summarizes it perfectly - we need best-effort, not an ideal.

Reply:
Thank you all for the feedback. 
I think a "best effort" is a fair description here.
I agree with the reasoning that we should not depend on `crio` or other things that are prone to fail the installation, leaving us as blind as we are today.

Therefore, this service needs to be basic and with little to no dependencies. Start via systemd as soon as boot happens and immediately try to report logs (and status).

I *think* this should cover both SNO and non-SNO cases; however, I am aware of some SNO-specific concerns @eranco74 has, and we will probably keep discussing that.

@romfreiman  as for:
>     2. no networking post reboot (static networking was not applied correctly) - low chances that despite that, the host will still have networking to reach the saas / hub cluster

as we have seen [today](https://coreos.slack.com/archives/CUPJTHQ5P/p1664781693190949?thread_ts=1664774218.591499&cid=CUPJTHQ5P) numbers do show that some clusters are using dhcp alongside nmstate, so we might also get logs in those situations.